### PR TITLE
Mitigate race conditions around enqueuing backgroundables + fix race conditions in traverse

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/FastForwardMergeBackgroundable.java [hunk 1]
```diff
@@ -2,28 +2,28 @@
...
-import lombok.experimental.ExtensionMethod;
 import lombok.val;
 import org.checkerframework.checker.guieffect.qual.UIEffect;
 import org.checkerframework.checker.tainting.qual.Untainted;
 
 import com.virtuslab.gitmachete.frontend.actions.common.MergeProps;
-import com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle;
+import com.virtuslab.qual.async.ContinuesInBackground;
 
-@ExtensionMethod(GitMacheteBundle.class)
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/FastForwardMergeBackgroundable.java:19]: Meh, some sh*t around Checker's not working here with `@ExtensionMethod`

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/ResetCurrentToRemoteBackgroundable.java [hunk 2]
```diff
@@ -18,14 +20,10 @@
...
 import git4idea.util.LocalChangesWouldBeOverwrittenHelper;
 import lombok.CustomLog;
-import lombok.experimental.ExtensionMethod;
 import lombok.val;
 
-import com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle;
-import com.virtuslab.gitmachete.frontend.vfsutils.GitVfsUtils;
 import com.virtuslab.qual.guieffect.UIThreadUnsafe;
 
-@ExtensionMethod({GitMacheteBundle.class, GitVfsUtils.class})
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/ResetCurrentToRemoteBackgroundable.java:28]: Again, Checker

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToRemote.java [hunk 3]
```diff
@@ -200,8 +203,15 @@ private boolean handleDivergedFromRemote(IManagedBranchSnapshot gitMacheteBranch
...
-        return true;
+            gitRepository) {
+          @Override
+          public void onSuccess() {
+            graphTable.queueRepositoryUpdateAndModelRefresh(traverseNextEntry);
+          }
+        }.queue();
+        // The ongoing traverse is now a responsibility of the freshly-queued backgroundable;
+        // NOT a responsibility of the outer method.
+        return false;
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToRemote.java:214]: Here's the actual fix...

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToRemote.java [hunk 4]
```diff
@@ -233,8 +244,13 @@ private boolean handleBehindRemote(IManagedBranchSnapshot gitMacheteBranch) {
...
         val fetchNotificationTextPrefix = fetchNotificationPrefix + (fetchNotificationPrefix.isEmpty() ? "" : " ");
-        new FastForwardMergeBackgroundable(gitRepository, mergeProps, fetchNotificationTextPrefix).queue();
-        return true;
+        new FastForwardMergeBackgroundable(gitRepository, mergeProps, fetchNotificationTextPrefix) {
+          @Override
+          public void onSuccess() {
+            graphTable.queueRepositoryUpdateAndModelRefresh(traverseNextEntry);
+          }
+        }.queue();
+        return false;
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToRemote.java:253]: ...and another one

File: qual/src/main/java/com/virtuslab/qual/async/ContinuesInBackground.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ * <li>methods that call {@code queue()} on classes extending {@code Task.Backgroundable} are marked as {@link ContinuesInBackground}</li>
+ * <li>methods that call {@link git4idea.branch.GitBrancher#checkout} are marked as {@link ContinuesInBackground}</li>
+ * <li>methods that call {@link ContinuesInBackground} methods are themselves marked as {@link ContinuesInBackground}</li>
+ * </ol>
+ * <p>
+ * This annotation must have runtime retention to be visible to ArchUnit tests.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD})
+public @interface ContinuesInBackground {}
```
Comment (PawelLipski) [inline qual/src/main/java/com/virtuslab/qual/async/ContinuesInBackground.java:26]: Would it prevent the race condition in question? Dunno, but at least it pokes the programmer when they try calling an "logically asynchronous" method from a "logically synchronous" one 🤔

File: src/test/java/com/virtuslab/archunit/BackgroundTaskEnqueuingTestSuite.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  private static List<String> extractWhitelistedMethodsFromAnnotation(JavaMethod method) {
+    if (method.isAnnotatedWith(IgnoreUIThreadUnsafeCalls.class)) {
+      return List.of(method.getAnnotationOfType(IgnoreUIThreadUnsafeCalls.class).value());
+    } else {
+      return List.empty();
+    }
+  }
+
+  @Test
+  public void only_continues_in_background_methods_should_call_other_continues_in_background_methods() {
```
Comment (PawelLipski) [inline src/test/java/com/virtuslab/archunit/BackgroundTaskEnqueuingTestSuite.java:37]: Mostly shamelessly copied from `UIThreadUnsafeMethodInvocationsTestSuite`... I'm sticking to _twice repeated? copy! thrice repeated? extract_ rule that I've learned in the olden days at Google ;p

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToRemote.java (hunk 3 above)
Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToRemote.java:214]: oh, I see. so previously reset could happen after the traverse of a next branch, right?

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToRemote.java [hunk 7]
```diff
@@ -233,8 +244,12 @@ private boolean handleBehindRemote(IManagedBranchSnapshot gitMacheteBranch) {
...
             : getNonHtmlString("action.GitMachete.BasePullAction.notification.prefix.fetch-perform");
         val fetchNotificationTextPrefix = fetchNotificationPrefix + (fetchNotificationPrefix.isEmpty() ? "" : " ");
-        new FastForwardMergeBackgroundable(gitRepository, mergeProps, fetchNotificationTextPrefix).queue();
-        return true;
+        Runnable doInUIThreadWhenReady = () -> graphTable.queueRepositoryUpdateAndModelRefresh(traverseNextEntry);
+        new FastForwardMergeBackgroundable(gitRepository, mergeProps, fetchNotificationTextPrefix, doInUIThreadWhenReady)
+            .queue();
+        // The ongoing traverse is now a responsibility of the freshly-queued backgroundable;
+        // NOT a responsibility of the outer method.
+        return false;
```
Comment (mkondratek) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToRemote.java:252]: and a similar issue here?

**Review by mkondratek — APPROVED** [review]

**Pull request merged**